tiny_cnn/layers/convolutional_layer.h

```diff
@@ -226,41 +225,43 @@ class convolutional_layer : public feedforward_layer<Activation> {
     }
 
     /**
-     * @param worker_index id of current worker-task
      * @param in_data      input vectors of this layer (data, weight, bias)
      * @param out_data     output vectors
      **/
-    void forward_propagation(cnn_size_t                 worker_index,
-                             const std::vector<vec_t*>& in_data,
-                             std::vector<vec_t*>&       out_data) {
+    void forward_propagation(const std::vector<tensor_t*>& in_data,
+                             std::vector<tensor_t*>&       out_data) {
         // launch convolutional kernel
-        Base::backend_->conv2d(worker_index, in_data, out_data);
+        Base::backend_->conv2d(in_data, out_data);
 
         // activations
-        vec_t& out     = *out_data[0];
-        const vec_t& a = *out_data[1];
+        for_i(in_data[0]->size(), [&](int sample) {
+            vec_t& out     = (*out_data[0])[sample];
+            const vec_t& a = (*out_data[1])[sample];
 
-        for_i(this->get_parallelize(), params_.out.size(), [&](int i) {
-            out[i] = this->h_.f(a, i);
+            for (cnn_size_t i = 0; i < params_.out.size(); i++) {
+                out[i] = this->h_.f(a, i);
+            };
         });
     }
 
     /**
      * return delta of previous layer (delta=\frac{dE}{da}, a=wx in fully-connected layer)
-     * @param worker_index id of current worker-task
      * @param in_data      input vectors (same vectors as forward_propagation)
      * @param out_data     output vectors (same vectors as forward_propagation)
      * @param out_grad     gradient of output vectors (i-th vector correspond with out_data[i])
      * @param in_grad      gradient of input vectors (i-th vector correspond with in_data[i])
      **/
-    void back_propagation(cnn_size_t                 worker_index,
-                          const std::vector<vec_t*>& in_data,
-                          const std::vector<vec_t*>& out_data,
-                          std::vector<vec_t*>&       out_grad,
-                          std::vector<vec_t*>&       in_grad) {
-        Base::backend_->conv2d(worker_index, in_data,
-                               out_data, out_grad, in_grad);
-      }
+    void back_propagation(const std::vector<tensor_t*>& in_data,
+                          const std::vector<tensor_t*>& out_data,
+                          std::vector<tensor_t*>&       out_grad,
+                          std::vector<tensor_t*>&       in_grad) {
+        Base::backend_->conv2d(in_data, out_data, out_grad, in_grad);
+    }
+
+    float_t& weight_at(cnn_size_t in_channel, cnn_size_t out_channel, cnn_size_t kernel_x, cnn_size_t kernel_y) {
+        vec_t* W = this->get_weights()[0];
+        return W[weight_.get_index(kernel_x, kernel_y, in_.depth_ * out_channel + in_channel)];
+    }
 
```

Comment:
@edgarriba 
Thanks, I got it. I mainly use a Github client and Visual Studio and sometimes use WinMerge. Do you know awsome tools for merging?